File: packages/core/src/detect.ts
import { basename, extname } from "node:path";
import { stat } from "node:fs/promises";
import type { Detection } from "./types.ts";
import { isDocumentPath, normalizeItemPath } from "./paths.ts";
import { textTitle } from "./text.ts";

export type PathProbe = (path: string) => Promise<"dir" | "file" | null>;

export const fsProbe: PathProbe = async path => {
  const info = await stat(path).catch(() => null);
  return info ? (info.isDirectory() ? "dir" : "file") : null;
};

/** A single-line http(s) URL: a PR when it points at a pull or merge request, otherwise a link. */
export function detectUrl(line: string): Detection | null {
  if (!/^https?:\/\/\S+$/.test(line)) return null;
  const github = line.match(/github\.com\/([^/]+\/[^/]+)\/pull\/(\d+)/);
  const generic = line.match(/\/(?:pull|merge_requests)\/(\d+)/);
  const url = new URL(line);
  if (github) return { shape: "url", type: "pr", url: line, title: `${github[1]}#${github[2]}` };
  if (generic) return { shape: "url", type: "pr", url: line, title: `${url.hostname}${url.pathname}` };
  return { shape: "url", type: "link", url: line, title: (url.hostname + url.pathname).replace(/\/$/, "") };
}

/** A single-line absolute or ~/ path: dir when it is one, document for .md/.html, else file. */
export async function detectPath(line: string, probe: PathProbe = fsProbe, home?: string): Promise<Detection | null> {
  if (!/^(~\/|\/|~$)/.test(line)) return null;
  const path = normalizeItemPath(line, home);
  const kind = await probe(path);
  const importable = isDocumentPath(path);
  const type = kind === "dir" ? "dir" : importable ? "document" : "file";
  const title = type === "document" ? basename(path, extname(path)) : basename(path);
  return { shape: "path", type, path, exists: kind !== null, title };
}

export const detectText = (content: string): Detection => ({ shape: "text", type: "note", content, title: textTitle(content) });

/**
 * What a pasted or picked thing is. `shape` says which field the text fills;
 * `type` is the guess a form may override.
 *   url  -> link, or pr when the URL points at a pull/merge request
 *   path -> dir when it is one on disk, document for .md/.html, else file
 *   text -> note (a note that merely contains a URL stays a note)
 */
export async function detect(raw: string, options: { probe?: PathProbe; home?: string } = {}): Promise<Detection> {
  const content = raw.trim();
  const line = content.split("\n").length === 1 ? content : "";
  if (line) {
    const url = detectUrl(line);
    if (url) return url;
    const path = await detectPath(line, options.probe, options.home);
    if (path) return path;
  }
  return detectText(content);
}

File: lib/pf/detect.rb
# frozen_string_literal: true

require 'pathname'
require 'uri'

module Pf
  DOCUMENT_EXTENSIONS = %w[.md .markdown .html].freeze
  MARKDOWN_EXTENSIONS = %w[.md .markdown].freeze

  module Detect
    module_function

    def expand_path(value, home = nil)
      home ||= ENV.fetch('HOME', '')
      return home if value == '~'
      return File.expand_path(value[2..], home) if value.start_with?('~/')

      value
    end

    def abbreviate_path(path, home = nil)
      home ||= ENV.fetch('HOME', '')
      home != '' && path.start_with?(home + '/') ? "~#{path[home.length..]}" : path
    end

    def normalize_item_path(value, home = nil)
      path = expand_path(value.to_s.strip, home)
      raise Pf::Error, 'path must be absolute or start with ~/' if path.empty? || !Pathname.new(path).absolute?

      File.expand_path(path)
    end

    def markdown_path?(path)
      MARKDOWN_EXTENSIONS.include?(File.extname(path).downcase)
    end

    def html_path?(path)
      File.extname(path).downcase == '.html'
    end

    def document_path?(path)
      DOCUMENT_EXTENSIONS.include?(File.extname(path).downcase)
    end

    def first_heading(markdown)
      markdown.to_s.each_line do |line|
        match = line.match(/^#\s+(.+)$/)
        return match[1].strip if match
      end
      nil
    end

    def first_line(text)
      text.to_s.each_line.map(&:strip).find { |line| !line.empty? }
    end

    def html_title(html)
      html.to_s.match(%r{<title[^>]*>([^<]+)</title>}i)&.[](1)&.strip
    end

    def document_title(content, source_path)
      fallback = File.basename(source_path, File.extname(source_path))
      return html_title(content) || fallback if html_path?(source_path)

      first_heading(content) || fallback
    end

    def text_title(content, max = 80)
      (first_heading(content) || first_line(content) || 'Untitled')[0, max]
    end

    def detect_url(line)
      return nil unless line.match?(%r{\Ahttps?://\S+\z})

      uri = URI.parse(line)
      raise URI::InvalidURIError, line unless uri.host

      host = uri.host.downcase
      github = line.match(%r{github\.com/([^/]+/[^/]+)/pull/(\d+)}i)
      generic = line.match(%r{/(?:pull|merge_requests)/(\d+)})
      if github
        { 'shape' => 'url', 'type' => 'pr', 'url' => line, 'title' => "#{github[1]}##{github[2]}" }
      elsif generic
        { 'shape' => 'url', 'type' => 'pr', 'url' => line, 'title' => "#{host}#{uri.path}" }
      else
        { 'shape' => 'url', 'type' => 'link', 'url' => line, 'title' => "#{host}#{uri.path}".sub(%r{/\z}, '') }
      end
    end

    def detect_path(line, home: nil)
      return nil unless line.match?(%r{\A(~/|/|~\z)})

      path = normalize_item_path(line, home)
      kind = if File.directory?(path)
               'dir'
             elsif File.exist?(path)
               'file'
             end
      type = kind == 'dir' ? 'dir' : (document_path?(path) ? 'document' : 'file')
      title = type == 'document' ? File.basename(path, File.extname(path)) : File.basename(path)
      { 'shape' => 'path', 'type' => type, 'path' => path, 'exists' => !kind.nil?, 'title' => title }
    end

    def detect_text(content)
      { 'shape' => 'text', 'type' => 'note', 'content' => content, 'title' => text_title(content) }
    end

    def detect(raw, home: nil)
      content = raw.to_s.strip
      line = content.include?("\n") ? '' : content
      if !line.empty?
        detected_url = detect_url(line)
        return detected_url if detected_url

        detected_path = detect_path(line, home: home)
        return detected_path if detected_path
      end
      detect_text(content)
    end

    def local_markdown_link(source_path, target)
      return nil if target.nil? || target.empty? || target.start_with?('#') || target.start_with?('//') || target.match?(%r{\A[a-z][a-z0-9+.-]*:}i)

      boundary = target.index(/[?#]/)
      encoded_path = boundary ? target[0...boundary] : target
      return nil if encoded_path.empty?

      begin
        path = URI::DEFAULT_PARSER.unescape(encoded_path)
      rescue ArgumentError, URI::InvalidURIError
        return nil
      end
      return nil unless markdown_path?(path)

      suffix = boundary ? target[boundary..] : ''
      { 'sourcePath' => File.expand_path(path, File.dirname(source_path)), 'suffix' => suffix }
    end
  end

end
